### DISCUSSION TITLE
Ideas - Improvements - Features

### DISCUSSION THREAD

**Discussion by bonzo81** [bonzo81/netbox-librenms-plugin#1] [Ideas]
Share any ideas for a future plugin feature here...

...or even better, send a PR! 😃

**Comment by donmart85** [comment]
Hi bonzo81,
 
first of all compliment, this is a beautiful plugin. I have installed it with NetBox Community v4.1.3 and works perfectly.
At the begin i have a problem with the start of plugin, i have replicated the configuration of plugin as written in the homepage, but it didn't work. 
i solved adding netbox_librenms_plugin and not netbox-librenms-plugin in configuration.py

PLUGINS = [
    'netbox_librenms_plugin'
]

PLUGINS_CONFIG = {
    'netbox_librenms_plugin': {
        'librenms_url': 'https://librenm.xxxxxxx.it/',
        'api_token': 'xxxxxxxxxxxxxxxxxxxxxxxxxxxx',
        'cache_timeout': 300,
    }
}

I have a lot of suggestions, i'm using both systems 

i begin with these:
1) Sync mac address with interface. 
2) Insert name of interface in Label and not in Description field.
3) Sync virtual-machine interface with same menu of devices. I have tried to modify your script but i am beginnnig to learn python and i am not able.
4) Select all interface for adding total interface not work.

Thanks. for your work.

**Reply by bonzo81** [reply]
Hi @donmart85 ,

Many thanks for your comments, and congrats on being first here! Really glad you've found it useful already.

I'll double check the configuration.py line, and update the README. Thanks for pointing that out.

1. Adding the Mac address to the interface sync should be straight forward to do. 

2. Could you give me an example in your setup? If there is no port description set in LibreNMS, it puts the port name in the ifAlias field . The ifAlias field is synced to the netbox interface description at the moment. My company uses port descriptions for describing the purpose of the interface, and info on the remote device. I think the label field in netbox is for alternative physical labels for when the physical port has a different label to the cli name. E.g. eth0 (cli) v 1 (physical). 

3. I'm not a virtual machine user but I think this would be possible. The same LibreNMS Sync tab for virtual machines?

4. Yes. I noticed this myself and will be fixed.

Keep going with python. I'm no developer so it's good to get feedback.

I will create issues for the items I can add sooner and tag you in them.

**Reply by donmart85** [reply]
Thank you for update and for your help. 

For Description, i use it for describing the purpose of the interface as you.
For example in linux server the interface's name was the same of description.
If it would be possible, it shouldn't write description for device interface's name if it was the same.
If i sync interface from LibreNMS, this action deletes my description.

For Virtual Machine, i am monitoring vm in Librenms as a normal device, but in Netbox i have inserted these object in section virtualization/virtual machine. 
I think that the import of interfaces should be the same of devices, with similar tab. 
The only difference is that in virtualization we haven't Interface Type Mappings and speed. 
Also in this case, it would be useful to have the mac-address's syncronization. 
I use Docker container. Every time i rebuild a container, the interface's name and mac address change.

The top would be an automatic update of all info with a script or similar system, but it's already excellent.

Sorry for my bad english XD
Have a nice day.

**Reply by bonzo81** [reply]
Grazie per il tuo contributo. Il tuo inglese è molto buono. (Uso ChatGPT per tradurre 😀). Posso guardare questo meglio la prossima settimana.

**Reply by donmart85** [reply]
Hi,
for point 3 i have created a new class of elements for Virtual Machine tab, it seems work correctly. Could you check it?
Thanks.
[netbox file mod.zip](https://github.com/user-attachments/files/17382482/netbox.file.mod.zip)

**Reply by bonzo81** [reply]
Hi @donmart85 ,

Great! Thank you for your continued contributions.

I have created an issue here - [Sync interfaces from virtual devices #11](https://github.com/bonzo81/netbox-librenms-plugin/issues/11)

If you feel able, please add your comments and especially your code via Git/Github. Hopefully the [contributing](https://github.com/bonzo81/netbox-librenms-plugin/blob/master/docs/contributing.md) page will help you get started. If not, please ask.

Using Github to track code changes makes the reviewing of code easier, and would allow you to contribute directly. Learning Git along with Python is a great idea. If you use VScode it can help make this process easier.

Thanks again.

**Comment by tkspuk** [comment]
Thank you for your great project. I tested it in Docker, and it worked successfully. I'm not sure what your future development plans are, but I have some ideas and wonder if they are on your roadmap:

Could it be possible to automatically add monitoring in LibreNMS based on specific NetBox attributes (such as role, status, and site) to avoid missing monitoring for newly added devices?
Synchronize more data, including LLDP connection information and more interface attributes.
Similar to the above, the bulk addition of all interfaces is currently not working.
Looking forward to continuous updates of the plugin!

**Reply by bonzo81** [reply]
Hi tkspuk,

Thanks for joining the discussion. Glad you have had success with the plugin already.

* There is definitely scope for automatic background syncing. However, I don't want to duplicate something already in Netbox like the use of webhooks to trigger adding devices to LibreNMS. Unless, it's an improved workflow. I plan to get the manual sync features well established before looking at what could be run automatically or as a scheduled task.

* I'm interested in your LLDP connection info idea. Are you thinking along the lines of creating a cable connection in Netbox from the LibreNMS  data? For example is it the same data as this web interface form in  Netbox?
![image](https://github.com/user-attachments/assets/ec553a03-9ddb-419a-93a3-d8f3038badbc)

* I've added an issue for the bulk addition [#4 here](https://github.com/bonzo81/netbox-librenms-plugin/issues/4). Hopefully get this added to the next release.

Lots of work to do!

**Comment by tkspuk** [comment]
Another issue I encountered is that some Ethernet interfaces are recognized as ethernetCsmacd after being synchronized to NetBox. This causes difficulties when mapping, as both physical 25G and 100G ports are identified the same way. Although I can differentiate them based on the synchronized speed, the mapping process is still a bit troublesome.

**Reply by bonzo81** [reply]
Can you provide more info? Did you have the interface mapping created for ethernetCsmacd? Did this mapping show correctly in the interface table before syncing? See [interface mapping docs page ](https://bonzo81.github.io/netbox-librenms-plugin/interface_mappings/) for more info.

**Reply by tkspuk** [reply]
![image](https://github.com/user-attachments/assets/e9e8a733-ea87-476e-9045-b4e43ce61c60)
For example, here both 100G and 25G ports are recognized as the same in LibreNMS.

**Reply by bonzo81** [reply]
It looks like the type mapping might need a speed value added for this scenario. I'll need to investigate this further. Thanks for the info

**Reply by bonzo81** [reply]
Hi @tkspuk,

I've created a Github issue for the interface type mapping issue you highlighted [here](https://github.com/bonzo81/netbox-librenms-plugin/issues/10).

Please feel free to add any other comments if helpful.

**Comment by kimarenchi** [comment]
I would suggest to have a device sync from LibreNMS to netbox with onboarding option from netbox

**Reply by bonzo81** [reply]
Hi @kimarenchi, thanks for your comment. Could you give more details about how this might work in the plugin?

**Reply by kimarenchi** [reply]
Hi @bonzo81, What I mean is to have an automated onboarding or adding of discovered devices from LibreNMS to Netbox.

Devices from LibreNMS, can have an onboarding tab and users can select which devices will be added to Netbox. This way we can still preserve the idea of netbox becoming the SOT as  users can still select which devices can be pulled from LibreNMS to netbox.

**Comment by donmart85** [comment]
Hi, thanks for your beautiful work, i told my ideas:

I would like that when I synchronize device's interfaces, the plugin allows me to choose to delete the interfaces present in Netbox but no longer present in LibreNMS. Example a changed interface's name in a Linux Server.

I have 2 LibreNMS servers, one for my corporation and one for customers. I would like to choose LibreNMS server without modify configuration.py.

I would like that plugin syncs also information as Hardware, Serial Number, Platform, Operating System.

**Reply by bonzo81** [reply]
Hey @donmart85, 

Thanks for the feedback. I will definitely keep these in mind for future releases. I don't have as much time to develop this plugin as before so I'll need to prioritise what to focus on. How would you prioritise your 3 suggestions?

**Reply by donmart85** [reply]
Hi @bonzo81 
they are in order of priority for me. 
thanks for your help.

**Reply by bonzo81** [reply]
Hi @donmart85,

I managed to get some development time this week so version 0.3.14 just released.

 I've added a display showing interfaces present in Netbox but not present in LIbreNMS. Clicking on it allows you to delete them.

<img width="293" height="158" alt="image" src="https://github.com/user-attachments/assets/db7689f1-81c3-4421-bf23-eca6ee0fe2fe" />


Also added ability to add multiple LibreNMS servers in configuration.py but select the active one in the GUI. This requires a slightly different configuration.py settings. See the docs on [multi server configuration](https://across-the-wire.gitbook.io/netbox-librenms-plugin/usage_tips/multi_server_configuration).

<img width="925" height="789" alt="image" src="https://github.com/user-attachments/assets/22d70085-661f-4e6a-8bd9-a5224f4443e2" />

 Hopefully this work for you (and others!).

**Reply by donmart85** [reply]
Hi @bonzo81 , 
wow it's incredible :)
It works like a charm!

Only a problem, the template configuration give me an error at startup.

django.core.exceptions.ImproperlyConfigured: Plugin netbox_librenms_plugin requires 'librenms_url' to be present in the PLUGINS_CONFIG section of configuration.py.

daemon "python3 ./manage.py rqworker" (pid: 259) annihilated

I have solved adding 2 lines into your config for default server
After plugin works

PLUGINS_CONFIG = {
    'netbox_librenms_plugin': {
        'servers': {
            'production': {
                'display_name': 'Production LibreNMS',
                'librenms_url': 'https://librenms-prod.example.com',
                'api_token': 'your_production_token',
                'cache_timeout': 300,
                'verify_ssl': True,
                'interface_name_field': 'ifDescr'
            },
            'testing': {
                'display_name': 'Test LibreNMS',
                'librenms_url': 'https://librenms-test.example.com',
                'api_token': 'your_test_token',
                'cache_timeout': 300,
                'verify_ssl': False,
                'interface_name_field': 'ifName'
            },
            'development': {
                'display_name': 'Dev LibreNMS',
                'librenms_url': 'https://librenms-dev.example.com',
                'api_token': 'your_dev_token',
                'cache_timeout': 180,
                'verify_ssl': False,
                'interface_name_field': 'ifDescr'
            }
        },
        "librenms_url": "https://librenms-prod.example.com",
        "api_token": "your_production_token"
    }
}

Best Regards.

**Reply by bonzo81** [reply]
Hi @donmart85 ,

Glad to hear it's working. 

You are right about the config problem. I will update this in the next release so that 'librenms_url' and 'api_token' are not required if the 'servers' block exists. 

Thanks for flagging that up.

**Comment by olerth** [comment]
Hi,
First of all - great plugin!
Second, a couple of minor requests.

1. Could we get SNMPv1 support? We still have devices using version 1 and it would be great to have them synced using this plugin.
2. In version 0.4 of the plugin, sync to LibreNMS no longer keeps previous values for auth username and such in memory. Would love if this could be an option to turn on/off!

**Reply by bonzo81** [reply]
Thanks @olerth, 
I will restore the auto complete field for the Add to Librenms form. Thanks for pointing that out.

Regarding SNMPv1 - I don't have any devices to check running snmpv1. Have you tried using the SNMPv2 form for a SNMP v1 device? I think Librenms might try to add a device in the order v2, v3 and then v1. SNMP v2 and v1 use the same parameters so it might work. If you are able to test this, and let me know that would be helpful. If this doesn't work I can add SNMP v1 to the form.

**Reply by olerth** [reply]
Hi @bonzo81,
I just tried again to verify that it is indeed not working for SNMPv1. Shouldn't be too difficult, for snmpwalk, the only difference is -v1 instead of -v2c

**Reply by bonzo81** [reply]
Thanks for checking @olerth. I will get v1 added to the 'Add to Librenms' form for next release
